### PR TITLE
Implement MediaStreamAudioDestinationNode, MediaStreamAudioSourceNode, MediaStreamTrackAudioSourceNode

### PR DESCRIPTION
**Pull request by Manishearth**
Progress in https://github.com/servo/servo/issues/26097

This is a draft since we need the data channels stuff to land first

(also I need to make sure we're passing WPT)

### CODE REVIEW

File: components/script/dom/mediastreamaudiosourcenode.rs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        let node_options = UnwrappedAudioNodeOptions {
+            count: 2,
+            mode: ChannelCountMode::Max,
+            interpretation: ChannelInterpretation::Speakers,
+        };
+        let node = AudioNode::new_inherited(
+            AudioNodeInit::MediaStreamSourceNode(track),
+            &context.upcast(),
+            node_options,
+            1, // inputs
```
Comment (ferjm) [inline components/script/dom/mediastreamaudiosourcenode.rs:51]: Likewise, as a source node, this should probably have 0 inputs.

File: components/script/dom/mediastreamaudiosourcenode.rs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            .get_tracks()
+            .iter()
+            .find(|t| t.ty() == MediaStreamType::Audio)
+            .ok_or(Error::InvalidState)?
+            .id();
+        let node_options = UnwrappedAudioNodeOptions {
+            count: 2,
+            mode: ChannelCountMode::Max,
+            interpretation: ChannelInterpretation::Speakers,
+        };
```
Comment (ferjm) [inline components/script/dom/mediastreamaudiosourcenode.rs:46]: These are the default values, so we can do `Default::default()` here.

File: components/script/dom/mediastreamtrackaudiosourcenode.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    pub fn new_inherited(
+        context: &AudioContext,
+        options: &MediaStreamTrackAudioSourceOptions,
+    ) -> Fallible<MediaStreamTrackAudioSourceNode> {
+        let track = options.mediaStreamTrack.id();
+        let node_options = UnwrappedAudioNodeOptions {
+            count: 2,
+            mode: ChannelCountMode::Max,
+            interpretation: ChannelInterpretation::Speakers,
+        };
```
Comment (ferjm) [inline components/script/dom/mediastreamtrackaudiosourcenode.rs:37]: Use `Default::default()` here, please.

**Review by ferjm — APPROVED** [review]
Looks good! Only a few minor issues to fix.

The data channels PR landed a couple of days ago. So this should be good to land after a rebase

**Comment by jdm** [comment]
```
./tests/html/media_stream_audio_nodes_loopback.html:12: no newline at EOF
```
Sigh.

**Comment by Manishearth** [comment]
@bors retry
 - network error on rustfmt

**Comment by Manishearth** [comment]
@bors-servo retry

spending too much time in rust land

**Comment by CYBAI** [comment]
```
  ▶ Unexpected subtest result in /webaudio/idlharness.https.window.html:
  └ PASS [expected FAIL] AudioContext interface: operation createMediaStreamSource(MediaStream)
  ▶ Unexpected subtest result in /webaudio/idlharness.https.window.html:
  └ PASS [expected FAIL] AudioContext interface: operation createMediaStreamTrackSource(MediaStreamTrack)
  ▶ Unexpected subtest result in /webaudio/idlharness.https.window.html:
  └ PASS [expected FAIL] AudioContext interface: operation createMediaStreamDestination()
  ▶ Unexpected subtest result in /webaudio/idlharness.https.window.html:
  └ PASS [expected FAIL] AudioContext interface: context must inherit property "createMediaStreamSource(MediaStream)" with the proper type
  ▶ Unexpected subtest result in /webaudio/idlharness.https.window.html:
  └ PASS [expected FAIL] AudioContext interface: calling createMediaStreamSource(MediaStream) on context with too few arguments must throw TypeError
  ▶ Unexpected subtest result in /webaudio/idlharness.https.window.html:
  └ PASS [expected FAIL] AudioContext interface: context must inherit property "createMediaStreamTrackSource(MediaStreamTrack)" with the proper type
  ▶ Unexpected subtest result in /webaudio/idlharness.https.window.html:
  └ PASS [expected FAIL] AudioContext interface: calling createMediaStreamTrackSource(MediaStreamTrack) on context with too few arguments must throw TypeError
  ▶ Unexpected subtest result in /webaudio/idlharness.https.window.html:
  └ PASS [expected FAIL] AudioContext interface: context must inherit property "createMediaStreamDestination()" with the proper type
  ▶ TIMEOUT [expected OK] /webaudio/the-audio-api/the-mediastreamaudiosourcenode-interface/mediastreamaudiosourcenode-routing.html
  │ 
  │ error: XDG_RUNTIME_DIR not set in the environment.
  │ libEGL warning: No hardware driver found, falling back to software rendering
  │ 0:00:00.497616481 10581 0x7f798c013d00 ERROR             jackclient gstjackaudioclient.c:35:jack_log_error: Cannot connect to server socket err = No such file or directory
  │ 0:00:00.497662445 10581 0x7f798c013d00 ERROR             jackclient gstjackaudioclient.c:35:jack_log_error: Cannot connect to server request channel
  │ 0:00:00.499792390 10581 0x7f798c013d00 ERROR             jackclient gstjackaudioclient.c:35:jack_log_error: jack server is not running or cannot be started
  │ 0:00:00.500007253 10581 0x7f798c013d00 ERROR             jackclient gstjackaudioclient.c:35:jack_log_error: JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
  │ 0:00:00.500015279 10581 0x7f798c013d00 ERROR             jackclient gstjackaudioclient.c:35:jack_log_error: JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
  │ 0:00:00.500020264 10581 0x7f798c013d00 WARN              jackclient gstjackaudioclient.c:381:gst_jack_audio_get_connection: could not create connection
  │ 0:00:00.500068384 10581 0x7f798c013d00 WARN                jacksink gstjackaudiosink.c:355:gst_jack_ring_buffer_open_device:<autoaudiosink0-actual-sink-jackaudio> error: Jack server not found
  │ 0:00:00.500074697 10581 0x7f798c013d00 WARN                jacksink gstjackaudiosink.c:355:gst_jack_ring_buffer_open_device:<autoaudiosink0-actual-sink-jackaudio> error: Cannot connect to the Jack server (status 17)
  │ 0:00:00.500697643 10581 0x7f798c013d00 WARN                 default oss4-property-probe.c:302:gst_oss4_property_probe_get_values:<autoaudiosink0-actual-sink-oss4> Can't open file descriptor to probe available devices: No such file or directory
  │ 0:00:00.500721718 10581 0x7f798c013d00 WARN                oss4sink oss4-sink.c:513:gst_oss4_sink_open:<autoaudiosink0-actual-sink-oss4> error: Could not open audio device for playback.
  │ 0:00:00.500727445 10581 0x7f798c013d00 WARN                oss4sink oss4-sink.c:513:gst_oss4_sink_open:<autoaudiosink0-actual-sink-oss4> error: system error: No such file or directory
  │ ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
  │ ALSA lib conf.c:4568:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
  │ ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
  │ ALSA lib conf.c:4568:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
  │ ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
  │ ALSA lib conf.c:4568:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
  │ ALSA lib conf.c:5047:(snd_config_expand) Evaluate error: No such file or directory
  │ ALSA lib pcm.c:2564:(snd_pcm_open_noupdate) Unknown PCM default
  │ AL lib: (EE) ALCplaybackAlsa_open: Could not open playback device 'default': No such file or directory
  │ 0:00:00.511310244 10581 0x7f798c013d00 WARN                  openal gstopenalsink.c:634:gst_openal_sink_open:<autoaudiosink0-actual-sink-openal> error: Could not open device.
  │ 0:00:00.511323336 10581 0x7f798c013d00 WARN                  openal gstopenalsink.c:634:gst_openal_sink_open:<autoaudiosink0-actual-sink-openal> error: ALC error: Out of Memory
  │ 0:00:00.511859379 10581 0x7f798c013d00 WARN                     oss gstosssink.c:398:gst_oss_sink_open:<autoaudiosink0-actual-sink-oss> error: Could not open audio device for playback.
  │ 0:00:00.511868946 10581 0x7f798c013d00 WARN                     oss gstosssink.c:398:gst_oss_sink_open:<autoaudiosink0-actual-sink-oss> error: system error: No such file or directory
  │ 0:00:00.512462260 10581 0x7f798c006d40 FIXME                default gstutils.c:3980:gst_pad_create_stream_id_internal:<appsrc0:src> Creating random stream-id, consider implementing a deterministic way of creating a stream-id
  │ 0:00:00.513797687 10581 0x7f798c046850 FIXME                default gstutils.c:3980:gst_pad_create_stream_id_internal:<appsrc1:src> Creating random stream-id, consider implementing a deterministic way of creating a stream-id
  └ 0:00:00.514872193 10581 0x7f79f811fde0 FIXME                default gstutils.c:3980:gst_pad_create_stream_id_internal:<appsrc2:src> Creating random stream-id, consider implementing a deterministic way of creating a stream-id
  ▶ Unexpected subtest result in /webaudio/the-audio-api/the-mediastreamaudiosourcenode-interface/mediastreamaudiosourcenode-routing.html:
  │ TIMEOUT [expected FAIL] MediaStreamAudioSourceNode captures the right track.
  └   → Test timed out
  ▶ Unexpected subtest result in /webaudio/the-audio-api/the-audionode-interface/audionode-connect-method-chaining.html:
  └ PASS [expected FAIL] Executing "media-group"
```

**Comment by Manishearth** [comment]
@bors-servo r=ferjm

Missed some changes in the rebasing.

The routing timeout is due to a bug in the analyzer node, not a bug in the media node

**Pull request merged**